src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,118 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] storage = (T[]) new Object[DEFAULT_CAPACITY];
+    private int size = 0;
+
     @Override
     public void add(T value) {
-
+        if (size == storage.length) {
+            T[] newStorage = (T[]) new Object[storage.length + (storage.length >> 1)];
+            for (int i = 0; i < storage.length; i++) {
+                newStorage[i] = storage[i];
+            }
+            storage = newStorage;
+        }
+        storage[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of array size");
+        }
```

Comment:
Create separate private method for this check because it is repeated in multiple places. Change it in all places